src/client/mod.rs

```diff
@@ -141,30 +146,65 @@ impl APIClient {
         }
     }
 
-    pub async fn request_events<T>(&self, request: http::Request<Vec<u8>>) -> Result<Vec<T>>
+    pub async fn request_events<T>(
+        &self,
+        request: http::Request<Vec<u8>>,
+    ) -> Result<impl Stream<Item = Result<T>>>
     where
         T: DeserializeOwned,
     {
-        let res : reqwest::Response = self.send(request).await?;
+        let res: reqwest::Response = self.send(request).await?;
         trace!("{} {}", res.status().as_str(), res.url());
-        //trace!("Response Headers: {:?}", res.headers());
-        let s = res.status();
-        let text = res.text().await?;
-        handle_api_errors(&text, &s)?;
 
-        // Should be able to coerce result into Vec<T> at this point
-        let mut xs : Vec<T> = vec![];
-        for l in text.lines() {
-            let r = serde_json::from_str(&l).map_err(|e| {
-                warn!("{} {:?}", l, e);
-                Error::SerdeError(e)
-            })?;
-            xs.push(r);
-        }
-        Ok(xs)
+        // Now use `unfold` to convert the chunked responses into a Stream
+        Ok(futures::stream::unfold(res, |mut resp| {
+            async move {
+                // A buffer to store our current object in
+                // as we stream it's entirety
+                let mut buff = Vec::new();
+
+                loop {
```

Comment:
Okay so I've added in some buffering, and this seems to work, I'm not running into any decode errors on the examples anymore. _However_ I really don't like how this solution has to do a `serde_json::from_slice` at each iterator just to see if the we need to keep buffering or not, that seems expensive. Not sure what else to do there though, there doesn't seem to be like a `serde_json::is_valid` method or something?

Reply:
I'm glad you pointed this out I totally wouldn't have noticed, that's a sneaky one haha

Reply:
Awesome. It seems to work now. Node informer is properly streaming in the example :tada:

Yeah, a full serde check every chunk seems expensive. The previous solution just used `.lines()` and that never ran into issues. Maybe we can just do a regex.

Reply:
Thinking about it some more; I was slightly surprised that it parsing per chunk works. That means that at least k8s ends the chunks on object end. That's fortunate. It means that it's probably okay to follow your approach. Maybe there are some official docs that specifies what we should do in this case :thinking:

Reply:
I ended up using the newline check like you suggested, I'm a bit surprised that it works, but it seems to be pretty stable. I can't find any explicit docs on the format of a kuberentes watch request, but playing around with `kubectl proxy` and something like `curl -s 127.0.0.1:8001/api/v1/watch/pods | jq --raw-output`, it's consistently events separated by new lines. 

It might be better to instead check if the current chunk _contains_ a new line rather than ends with it, to help catch overlapped chunks, but it seems like those might never be sent?

Reply:
The current solution is having problems for me. Posting a crash below from `node_informer`.

If we are going to split based on newlines we should probably test it out on a resource that inlines newlines, like `ConfigMap`.

Reply:
I'm thinking it's probably fine to operate under the assumption that a chunk end => object end for now - maybe we can ask maintainers if that's sensible. It certainly seems to hold for me.

Reply:
https://github.com/kubernetes/kubernetes/issues/24729

> We currently invoke flush on the HTTP response after every watch event, in order to guarantee clients see watches immediately.

Reply:
Although, it seems that they might not flush if they have more objects in the queue. https://github.com/kubernetes/kubernetes/issues/24729 did make it in.

Maybe we both just have underutilized api servers.

Reply:
Hmm that's problematic then, I've pushed some changes that use `IgnoredAny` from serde to just check valid JSON, which should be pretty fast and a bit more resilient than the newline check. Though if kubernetes is potentially pushing more than one event per request, and chunk end != object end, it almost feels like we'll need to wait for the whole request to finish before attempting to parse it?

Reply:
I think we can still do it speculatively, left a comment below. Somehow it got disconnected from this thread.